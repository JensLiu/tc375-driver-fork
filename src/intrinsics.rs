//! Intrinsic functions for the TriCore architecture.
//!
//! These functions are used to access the hardware directly and are
//! architecture specific.
//! When the tracing feature is enabled, these functions are replaced with
//! tracing functions that log the access to the hardware.

// TODO Remove once the unsafe blocks are documented
#![allow(clippy::undocumented_unsafe_blocks)]

#[cfg(feature = "tracing")]
use crate::tracing;

/// Atomic Load-Modify-Store, store under a `mask` of a `value` to the address `addr`.
/// This is needed to atomically update a memory location or to
/// track the changes to a memory location when tracing feature is enabled.
#[allow(unreachable_code)]
#[inline(always)]
pub(crate) unsafe fn load_modify_store(addr: *mut u32, v: u32, m: u32) {
    #[cfg(feature = "tracing")]
    {
        return tracing::load_modify_store(addr as usize, u64::from(v) | (u64::from(m) << 32));
    }

    #[cfg(target_arch = "tricore")]
    {
        return unsafe {
            core::arch::tricore::intrinsics::__ldmst(addr, v, m);
        };
    }

    panic!("unsupported architecture");
}

/// Volatile write to a memory location.
/// This is the equivalent of ptr.write_volatile(val) but it is tracked when the tracing feature is enabled.
#[allow(unreachable_code)]
#[allow(dead_code)]
#[inline(always)]
pub(crate) unsafe fn write_volatile<T>(addr: *mut T, val: T)
where
    u64: From<T>,
{
    #[cfg(feature = "tracing")]
    {
        return tracing::write_volatile(addr as usize, core::mem::size_of::<T>(), val.into());
    }

    #[cfg(target_arch = "tricore")]
    {
        return unsafe { addr.write_volatile(val) };
    }

    panic!("unsupported architecture");
}

/// Volatile read from a memory location.
/// This is the equivalent of ptr.read_volatile() but it is tracked when the tracing feature is enabled.
#[allow(unreachable_code)]
#[inline(always)]
pub(crate) unsafe fn read_volatile<T>(addr: *const T) -> T
where
    T: From<u32>,
{
    #[cfg(feature = "tracing")]
    {
        let val: u64 = tracing::read_volatile(addr as usize, core::mem::size_of::<T>());
        let val: u32 = val as u32;
        return val.into();
    }

    #[cfg(target_arch = "tricore")]
    {
        return unsafe { addr.read_volatile() };
    }

    panic!("unsupported architecture");
}

#[inline(always)]
pub(crate) fn read_cpu_core_id() -> u32 {
    #[cfg(feature = "tracing")]
    {
        0
    }

    #[cfg(target_arch = "tricore")]
    unsafe {
        let value: u32;
        core::arch::asm!("mfcr {0}, 0xFE1C", out(reg32) value);
        value
    }
}

#[cfg(feature = "tracing")]
#[cfg(test)]
mod tests {
    use super::*;
    use std::string::ToString;

    #[test]
    fn test_load_modify_store() {
        let report = tracing::log::Report::new();
        unsafe {
            load_modify_store(0x1000 as *mut u32, 0x1234, 0x5678);
        }
        assert_eq!(
            report.take_log().to_string(),
            "ldms 0x00001000 0x00005678 0x00001234\n"
        );
    }

    #[test]
    fn test_write_volatile() {
        let report = tracing::log::Report::new();
        let v_ptr = 0x1000 as *mut u32;
        unsafe {
            write_volatile(v_ptr, 0x5678);
        }
        assert_eq!(
            report.take_log().to_string(),
            "w    0x00001000 04 0x00005678\n"
        );
    }

    #[test]
    fn test_read_volatile() {
        let report = tracing::log::Report::new();
        let v_ptr = 0x1000 as *mut u32;
        report.expect_read(0x1000, 4, 0x1234);
        let val = unsafe { read_volatile(v_ptr) };
        assert_eq!(val, 0x1234);
        assert_eq!(
            report.take_log().to_string(),
            "r    0x00001000 04 0x00001234\n"
        );
    }
}
