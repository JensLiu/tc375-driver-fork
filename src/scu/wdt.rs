#![allow(clippy::cast_possible_truncation)]

use crate::intrinsics::read_cpu_core_id;
use crate::pac;

// If PAS=0: WDTxCON0.PW[7:2] must be written with inverted current value read from WDTxCON0.PW[7:2]
fn correct_password(password: u16) -> u16 {
    password ^ 0x003F
}

#[inline]
fn get_wdt_con0(core_id: u8) -> pac::Reg<pac::scu::wdtcpu::WdtcpUyCon0_SPEC, pac::RW> {
    pac::SCU.wdtcpu()[usize::from(core_id)].wdtcpuycon0()
}

#[inline]
fn get_wdt_con1(core_id: u8) -> pac::Reg<pac::scu::wdtcpu::WdtcpUyCon1_SPEC, pac::RW> {
    pac::SCU.wdtcpu()[usize::from(core_id)].wdtcpuycon1()
}

#[inline]
fn get_wdts_con0() -> pac::Reg<pac::scu::wdts::Wdtscon0_SPEC, pac::RW> {
    pac::SCU.wdts().wdtscon0()
}

#[inline]
fn get_wdts_con1() -> pac::Reg<pac::scu::wdts::Wdtscon1_SPEC, pac::RW> {
    pac::SCU.wdts().wdtscon1()
}

// TODO Duplicate? Bad function name?
#[inline]
pub(crate) fn clear_cpu_endinit_inline() {
    let core_id = read_cpu_core_id();
    let con0 = get_wdt_con0(core_id as u8);

    // SAFETY: core_id is read through assembly instruction MFCR
    let mut wdtcon0 = unsafe { con0.read() };
    // If PAS=0: WDTxCON0.PW[7:2] must be written with inverted current value read from WDTxCON0.PW[7:2]
    let password = correct_password(wdtcon0.pw().get());

    let rel = wdtcon0.rel().get();

    if wdtcon0.lck().get() {
        wdtcon0 = wdtcon0
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        // SAFETY: Each bit of WDTCPUyCON0 is at least W
        unsafe { con0.write(wdtcon0) };
    }

    wdtcon0 = wdtcon0
        .endinit()
        .set(false)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);
    // SAFETY: Each bit of WDTCPUyCON0 is at least W
    unsafe { con0.write(wdtcon0) };

    // TODO This conditional compilation can be removed, now that tracing can set the register value
    #[cfg(tricore_arch = "tricore")]
    while unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub(crate) fn set_cpu_endinit_inline() {
    let core_id = read_cpu_core_id();
    let con0 = get_wdt_con0(core_id as u8);
    // SAFETY: core_id is read through assembly instruction MFCR
    let mut wdtcon0 = unsafe { con0.read() };

    let password = correct_password(wdtcon0.pw().get());

    let rel = wdtcon0.rel().get();

    if wdtcon0.lck().get() {
        wdtcon0 = wdtcon0
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        // SAFETY: Each bit of WDTCPUyCON0 is at least W
        unsafe { con0.write(wdtcon0) };
    }

    wdtcon0 = wdtcon0
        .endinit()
        .set(true)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);

    // SAFETY: Each bit of WDTCPUyCON0 is at least W
    unsafe { con0.write(wdtcon0) };

    // TODO This conditional compilation can be removed, now that tracing can set the register value
    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub(crate) fn clear_safety_endinit_inline() {
    let con0 = get_wdts_con0();

    let mut wdtcon0 = unsafe { con0.read() };
    let password = correct_password(wdtcon0.pw().get());

    let rel = wdtcon0.rel().get();

    // SAFETY: LCK is a RWH bit
    if wdtcon0.lck().get() {
        wdtcon0 = wdtcon0
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(wdtcon0) };
    }

    wdtcon0 = wdtcon0
        .endinit()
        .set(false)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);

    // SAFETY: Each bit of WDTSCON0 is RW
    unsafe { con0.write(wdtcon0) };

    #[cfg(tricore_arch = "tricore")]
    while unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub(crate) fn set_safety_endinit_inline() {
    let con0 = get_wdts_con0();

    let mut wdtcon0 = unsafe { con0.read() };
    let password = correct_password(wdtcon0.pw().get());

    let rel = wdtcon0.rel().get();

    if wdtcon0.lck().get() {
        wdtcon0 = wdtcon0
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(wdtcon0) };
    }

    wdtcon0 = wdtcon0
        .endinit()
        .set(true)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);
    unsafe { con0.write(wdtcon0) };

    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

pub fn disable_safety_watchdog() {
    clear_safety_endinit_inline();

    let con1 = get_wdts_con1();
    // SAFETY: DR is a RW bit, it can be modified only when safety endinit is de-asserted (clear_safety_endinit_inline)
    unsafe { con1.modify(|p| p.dr().set(true)) };

    set_safety_endinit_inline();
}

pub fn disable_cpu_watchdog() {
    clear_cpu_endinit_inline();

    let core_id = read_cpu_core_id();
    // SAFETY: core_id is read through assembly instruction MFCR
    let con1 = get_wdt_con1(core_id as u8);
    // SAFETY: DR is a RW bit, it can be modified only when safety endinit is de-asserted (clear_cpu_endinit_inline)
    unsafe { con1.modify(|p| p.dr().set(true)) };

    set_cpu_endinit_inline();
}

#[cfg(feature = "tracing")]
#[cfg(test)]
mod tests {
    use super::*;
    use crate::tracing::log::Report;

    #[test]
    fn test_clear_cpu_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdt_con0(0);
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_cpu_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_set_cpu_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdt_con0(0);
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        set_cpu_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_clear_then_set_cpu_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdt_con0(0);
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_cpu_endinit_inline();
        report.expect_read(con0.ptr() as usize, 4, 0b10);
        set_cpu_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_set_then_clear_cpu_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdt_con0(0);
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        set_cpu_endinit_inline();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_cpu_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_clear_safety_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdts_con0();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_safety_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_set_safety_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdts_con0();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        set_safety_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_clear_then_set_safety_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdts_con0();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_safety_endinit_inline();
        report.expect_read(con0.ptr() as usize, 4, 0b10);
        set_safety_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_set_then_clear_safety_endinit_inline() {
        let report = Report::new();
        let con0 = get_wdts_con0();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        set_safety_endinit_inline();
        report.expect_read(con0.ptr() as usize, 4, 0b11);
        clear_safety_endinit_inline();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_disable_cpu_watchdog() {
        let report = Report::new();
        let con0 = get_wdt_con0(0);
        let con1 = get_wdt_con1(0);

        report.expect_read(con0.ptr() as usize, 4, 0b11);
        report.expect_read(con1.ptr() as usize, 4, 0b00);
        report.expect_read(con0.ptr() as usize, 4, 0b10);

        disable_cpu_watchdog();
        insta::assert_snapshot!(report.take_log());
    }

    #[test]
    fn test_disable_safety_watchdog() {
        let report = Report::new();
        let con0 = get_wdts_con0();
        let con1 = get_wdts_con1();

        report.expect_read(con0.ptr() as usize, 4, 0b11);
        report.expect_read(con1.ptr() as usize, 4, 0b00);
        report.expect_read(con0.ptr() as usize, 4, 0b10);

        disable_safety_watchdog();
        insta::assert_snapshot!(report.take_log());
    }
}
