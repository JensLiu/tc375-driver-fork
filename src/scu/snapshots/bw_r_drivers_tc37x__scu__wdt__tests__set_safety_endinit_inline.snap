---
source: src/scu/wdt.rs
assertion_line: 210
expression: report.take_log()
---
r    0xF00362A8 04 0x00000003
w    0xF00362A8 04 0x000000FD
w    0xF00362A8 04 0x000000FF
